ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs

```diff
@@ -742,27 +742,33 @@ withPeerStateActions PeerStateActionsArguments {
                       ))
 
 
-    -- 'monitorPeerConnection' is only used against established connections
+    -- 'monitorPeerConnection' is only used against established connections.
+    -- It returns 'Nothing' only if all mini-protocols are either not running
+    -- or still executing.
+    --
     monitorPeerConnection :: PeerConnectionHandle muxMode peerAddr ByteString m a b
-                          -> STM m (PeerStatus, ReconnectDelay)
+                          -> STM m (PeerStatus, Maybe ReconnectDelay)
     monitorPeerConnection PeerConnectionHandle { pchPeerStatus, pchAppHandles } =
         (,) <$> readTVar pchPeerStatus
             <*> (g <$> traverse f pchAppHandles)
       where
         f :: ApplicationHandle muxMode ByteString m a b
-          -> STM m (Map MiniProtocolNum (HasReturned a))
-        f = sequence <=< readTVar . ahMiniProtocolResults
+          -> STM m (Map MiniProtocolNum (Maybe (HasReturned a)))
+             -- do not block when a mini-protocol is still running
+        f =  traverse (((\stm -> (Just <$> stm) `orElse` pure Nothing)))
+         <=< readTVar . ahMiniProtocolResults
```

Comment:
I believe this is the culprit? Maybe explain here with more detail why we do not want this to block.

Reply:
Done